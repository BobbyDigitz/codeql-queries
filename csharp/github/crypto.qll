import csharp

module Crypto {
  class HashingAlgorithm extends DataFlow::ExprNode {
    abstract DataFlow::ExprNode getHashValue();

    abstract DataFlow::ExprNode getSalt();

    abstract DataFlow::ExprNode getIterations();
  }

  class SymmetricAlgorithm extends DataFlow::ExprNode {
    abstract int maxKeySize();

    abstract int minKeySize();

    abstract int getKeySize();
  }

  // Abstraction classes
  abstract class HashingAlgorithms extends HashingAlgorithm { }

  abstract class SymmetricAlgorithms extends SymmetricAlgorithm { }

  // Content
  class CryptoRfc2898DeriveBytes extends HashingAlgorithms {
    CryptoRfc2898DeriveBytes() {
      exists(ObjectCreation object |
        object.getType().getQualifiedName() = "System.Security.Cryptography.Rfc2898DeriveBytes" and
        this.asExpr() = object
      )
    }

    override DataFlow::ExprNode getHashValue() {
      result.asExpr() = this.asExpr().(ObjectCreation).getArgument(0)
    }

    override DataFlow::ExprNode getSalt() {
      result.asExpr() = this.asExpr().(ObjectCreation).getArgument(1)
    }

    override DataFlow::ExprNode getIterations() {
      result.asExpr() = this.asExpr().(ObjectCreation).getArgument(2)
    }
  }

  class DSACryptoServiceProvider extends SymmetricAlgorithms {
    DSACryptoServiceProvider() {
      exists(ObjectCreation object |
        object
            .getType()
            .hasQualifiedName("System.Security.Cryptography", "DSACryptoServiceProvider") and
        this.asExpr() = object
      )
    }

    override int maxKeySize() { result = 1024 }

    override int minKeySize() { result = 1024 }

    override int getKeySize() {
      this.asExpr().(ObjectCreation).hasNoArguments() and
      result = 1024
      or
      // https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.dsacryptoserviceprovider.-ctor?view=net-6.0#system-security-cryptography-dsacryptoserviceprovider-ctor(system-int32)
      result = this.asExpr().(ObjectCreation).getArgument(0).getValue().toInt()
    }
  }

  class RC2CryptoServiceProvider extends SymmetricAlgorithms {
    RC2CryptoServiceProvider() {
      exists(ObjectCreation object |
        object
            .getType()
            .hasQualifiedName("System.Security.Cryptography", "RC2CryptoServiceProvider") and
        this.asExpr() = object
      )
    }

    override int maxKeySize() { result = 128 }

    override int minKeySize() { result = 128 }

    override int getKeySize() {
      this.asExpr().(ObjectCreation).hasNoArguments() and
      result = 128 // default
      or
      // https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.dsacryptoserviceprovider.-ctor?view=net-6.0#system-security-cryptography-dsacryptoserviceprovider-ctor(system-int32)
      result = this.asExpr().(ObjectCreation).getArgument(0).getValue().toInt()
    }
  }

  class RSA extends SymmetricAlgorithms {
    RSA() {
      exists(ObjectCreation object |
        object
            .getType()
            .hasQualifiedName("System.Security.Cryptography", ["RSACryptoServiceProvider", "RSACng"]) and
        this.asExpr() = object
      )
      or
      exists(MethodCall call |
        call.getType().hasQualifiedName("System.Security.Cryptography", ["RSA"]) and
        call.getTarget().hasName("Create") and
        this.asExpr() = call
      )
    }

    override int maxKeySize() { result = 2048 }

    override int minKeySize() { result = 2048 }

    override int getKeySize() {
      (
        // https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.dsacryptoserviceprovider.-ctor?view=net-6.0#system-security-cryptography-dsacryptoserviceprovider-ctor(system-int32)
        this.asExpr().(ObjectCreation).hasNoArguments() and
        result = 1024
        or
        result = this.asExpr().(ObjectCreation).getArgument(0).getValue().toInt()
      )
      or
      (
        // https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.rsa.create?view=net-6.0#system-security-cryptography-rsa-create(system-int32)
        this.asExpr().(MethodCall).hasNoArguments() and
        result = 1024
        or
        result = this.asExpr().(MethodCall).getArgument(0).getValue().toInt()
      )
    }
  }
}
