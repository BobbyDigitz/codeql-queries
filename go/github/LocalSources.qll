import go

// ========== Sources ==========
module LocalSources {
  abstract class Sources extends DataFlow::Node { }

  // https://pkg.go.dev/os#Getenv
  class OsGetenv extends Sources, DataFlow::CallNode {
    OsGetenv() { this.getTarget().hasQualifiedName(package("os", ""), "Getenv") }
  }

  // https://pkg.go.dev/flag
  class Flag extends Sources, DataFlow::CallNode {
    Flag() { this.getTarget().hasQualifiedName(package("flag", ""), "String") }
  }
}
