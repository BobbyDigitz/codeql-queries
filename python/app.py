
from flask import Flask, render_template, request, redirect, url_for


app = Flask(__name__)


@app.route('/')
def index():
    name = request.args.get('name', 'World')
    return f'<h1>Hello {name}!</h1>'


if __name__ == "__main__":
    app.run(debug=True)
