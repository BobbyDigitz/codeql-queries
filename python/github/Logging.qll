import python
import semmle.python.ApiGraphs

abstract class LoggingSinks extends DataFlow::Node { }

class PrintMethod extends LoggingSinks {
  PrintMethod() {
    exists(DataFlow::Node call |
      call = API::builtin("print").getACall() and
      call.getScope().inSource() and
      this = call
    )
  }
}

class LoggingFramework extends LoggingSinks {
  LoggingFramework() {
    exists(DataFlow::Node call, e() and
      this = call
    )
  }
}
