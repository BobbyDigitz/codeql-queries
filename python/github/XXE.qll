import python
import semmle.python.dataflow.new.DataFlow
import semmle.python.dataflow.new.RemoteFlowSources
import semmle.python.dataflow.new.TaintTracking
import semmle.python.Concepts
import semmle.python.ApiGraphs

module XXE {
  abstract class Source extends DataFlow::Node { }

  abstract class Sink extends DataFlow::Node { }

  abstract class Sanitizer extends DataFlow::Node { }

  class RemoteFlowSourceAsSource extends Source, RemoteFlowSource { }

  API::Node getPyXMLParser() {
    // Find Insecure parsers
    exists(API::Node nodes, DataFlow::CallCfgNode feature |
      // > from xml.sax import make_parser
      // > parser = make_parser()
      nodes = API::moduleImport("xml.sax").getMember("make_parser") and
      // Make sure that the feature is enabled
      // > from xml.sax.handler import feature_external_ges
      // > parser.setFeature(feature_external_ges, True)
      feature = nodes.getReturn().getMember("setFeature").getACall() and
      feature.getArg(0) =
        API::moduleImport("xml.sax.handler").getMember("feature_external_ges").getAUse() and
      feature.getArg(1).asExpr().(BooleanLiteral).booleanValue() = true and
      result = nodes.getReturn()
    )
  }

  class PyXML extends Sink {
    PyXML() {
      exists(DataFlow::CallCfgNode call |
        // > from xml.dom.pulldom import parseString
        // > parseString(request.body.decode('utf-8'), parser=parser)
        call = API::moduleImport("xml.dom.pulldom").getMember("parseString").getACall() and
        call.getArgByName("parser") = getPyXMLParser().getAUse() and
        this = call.getArg(0)
      )
    }
  }
}
