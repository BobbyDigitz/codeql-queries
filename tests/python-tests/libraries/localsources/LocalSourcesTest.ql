import python
import github.LocalSources

from LocalSources sources
select sources
