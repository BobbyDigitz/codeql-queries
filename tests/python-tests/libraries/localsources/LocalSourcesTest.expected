| app.py:7:6:7:15 | ControlFlowNode for Attribute |
| app.py:8:6:8:15 | ControlFlowNode for Attribute |
| app.py:9:6:9:15 | ControlFlowNode for Attribute |
| app.py:12:6:12:13 | ControlFlowNode for Attribute |
| app.py:15:6:15:22 | ControlFlowNode for input() |
| app.py:20:8:20:26 | ControlFlowNode for Attribute() |
| app.py:22:6:22:11 | ControlFlowNode for Attribute |
| app.py:23:6:23:15 | ControlFlowNode for Attribute |
| app.py:26:5:26:23 | ControlFlowNode for open() |
| app.py:27:6:27:13 | ControlFlowNode for Attribute() |
| app.py:29:6:29:24 | ControlFlowNode for open() |
| app.py:30:10:30:17 | ControlFlowNode for Attribute() |
| app.py:32:6:32:40 | ControlFlowNode for Attribute() |
| app.py:33:7:33:23 | ControlFlowNode for Attribute() |
